examples/cis2-multi-royalties/src/lib.rs

```diff
@@ -0,0 +1,1547 @@
+                .balances
+                .entry(*token_id)
+                .occupied_or(ContractError::InsufficientFunds)?;
+            ensure!(*from_balance >= amount, ContractError::InsufficientFunds);
+            *from_balance -= amount;
+        }
+
+        if let Some(royalties) = royalties {
+            let mut royalties_state = self
+                .state
+                .entry(royalties.royalty_receiver)
+                .occupied_or(CustomContractError::InvalidRoyaltyAddress)?;
+
+            let mut minter_balance =
+                royalties_state.balances.entry(*token_id).or_insert(TokenAmountU64(0));
+
+            *minter_balance += concordium_cis2::TokenAmountU64(royalties.payment);
+            amount -= concordium_cis2::TokenAmountU64(royalties.payment);
+        }
+
+        let mut to_address_state =
+            self.state.entry(*to).or_insert_with(|| AddressState::empty(state_builder));
+        let mut to_address_balance = to_address_state.balances.entry(*token_id).or_insert(0.into());
+        *to_address_balance += amount;
+
+        Ok(())
+    }
+
+    /// Update the state adding a new operator for a given address.
+    /// Succeeds even if the `operator` is already an operator for the
+    /// `address`.
+    fn add_operator(
+        &mut self,
+        owner: &Address,
+        operator: &Address,
+        state_builder: &mut StateBuilder<S>,
+    ) {
+        let mut owner_state =
+            self.state.entry(*owner).or_insert_with(|| AddressState::empty(state_builder));
+        owner_state.operators.insert(*operator);
+    }
+
+    /// Update the state removing an operator for a given address.
+    /// Succeeds even if the `operator` is not an operator for the `address`.
+    fn remove_operator(&mut self, owner: &Address, operator: &Address) {
+        self.state.entry(*owner).and_modify(|address_state| {
+            address_state.operators.remove(operator);
+        });
+    }
+
+    /// Check if state contains any implementors for a given standard.
+    fn have_implementors(&self, std_id: &StandardIdentifierOwned) -> SupportResult {
+        if let Some(addresses) = self.implementors.get(std_id) {
+            SupportResult::SupportBy(addresses.to_vec())
+        } else {
+            SupportResult::NoSupport
+        }
+    }
+
+    /// Set implementors for a given standard.
+    fn set_implementors(
+        &mut self,
+        std_id: StandardIdentifierOwned,
+        implementors: Vec<ContractAddress>,
+    ) {
+        self.implementors.insert(std_id, implementors);
+    }
+
+    fn calculate_royalty_payments(
+        &mut self,
+        params: CheckRoyaltyParams,
+    ) -> Result<CheckRoyaltyResult, ContractError> {
+        let token_details =
+            self.token_details.get(&params.id).ok_or(ContractError::InvalidTokenId)?;
+
+        let minter = token_details.minter;
+        let royalty_to_pay: u64 = params.sale_price * token_details.royalty as u64 / 100;
```

Comment:
Thinking about it, we should probably prevent overflow here.

```suggestion
        let royalty_to_pay: u64 = params.sale_price.checked_mul(token_details.royalty as u64).ok_or(CustomContractError::Overflow)? / 100;
```
     
Adding an extra error to the `CustomContractError` struct.
``` 
 /// Overflow in royalty calculation.
Overflow,
```


https://doc.rust-lang.org/std/primitive.u64.html#method.checked_mul

Reply:
Good idea - done